### PR TITLE
introduce 0.7.0 entrypoint

### PR DESCRIPTION
**Pull request by zsluedem**

### CODE REVIEW

File: .github/scripts/install_solc.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@ set -eo pipefail
 
 # install solc
 
-SOLC_VERSION="v0.8.20"
+SOLC_VERSION="v0.8.23"
```
Comment (Vid201) [inline .github/scripts/install_solc.sh:7]: Update SOLC version in README

File: crates/primitives/src/utils.rs [hunk 2]
```diff
@@ -14,18 +15,15 @@ where
 }
 
 /// Converts bytes to checksum (first 20 bytes are address)
```
Comment (Vid201) [inline crates/primitives/src/utils.rs:17]: Update comment

File: crates/primitives/src/utils.rs [hunk 3]
```diff
@@ -52,3 +50,105 @@ pub fn get_address(buf: &[u8]) -> Option<Address> {
         None
     }
 }
+
+pub fn unpack_paymaster_fee_data(buf: &[u8]) -> (Address, U256, U256, Bytes) {
```
Comment (Vid201) [inline crates/primitives/src/utils.rs:54]: I think we can rename this to `unpack_paymaster_data`, and `unpack_factory_data` ...

File: crates/primitives/src/user_operation/mod.rs [hunk 4]
```diff
@@ -67,8 +69,9 @@ pub struct UserOperationSigned {
     /// Nonce (anti replay protection)
     pub nonce: U256,
 
-    /// Init code for the account (needed if account not yet deployed and needs to be created)
-    pub init_code: Bytes,
+    pub factory: Address,
```
Comment (Vid201) [inline crates/primitives/src/user_operation/mod.rs:72]: Add docs

File: crates/primitives/src/user_operation/mod.rs [hunk 5]
```diff
@@ -440,11 +470,71 @@ pub struct UserOperationReceipt {
     pub tx_receipt: TransactionReceipt,
 }
 
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct RpcUserOperation {
```
Comment (Vid201) [inline crates/primitives/src/user_operation/mod.rs:475]: Rename to `UserOperationRpc` ? I don't feel strongly about that, but other user operation names have suffix

File: crates/contracts/src/simulation_codes.rs [hunk 6]
```diff
@@ -0,0 +1 @@
...fffc803d016004833e81513d67ffffffffffffffff81602484011181841117156150ae57505050505090565b82850191508151818111156150c65750505050505090565b843d87010160208285010111156150e05750505050505090565b6150ef60208286010187614052565b509095945050505050565b60608152600061510d6060830186614c42565b60208301949094525060400152919050565b60006020828403121561513157600080fd5b5051919050565b82815260606020820152600d60608201527f4141323320726576657274656400000000000000000000000000000000000000608082015260a0604082015260006108af60a08301846146a3565b6000806040838503121561519857600080fd5b825167ffffffffffffffff8111156151af57600080fd5b8301601f810185136151c057600080fd5b80516151cb816140c7565b6040516151d88282614052565b8281528760208486010111156151ed57600080fd5b6151fe83602083016020870161467f565b6020969096015195979596505050505050565b82815260606020820152600d60608201527f4141333320726576657274656400000000000000000000000000000000000000608082015260a0604082015260006108af60a08301846146a3565b7fffffffffffffffffffffffffffffffff000000000000000000000000000000008135818116916010851015614a355760109490940360031b84901b169092169291505056fea2646970667358221220da6235a9fed490e0598819f695bb128f935391fa9c8ba963180dfb5cab452aef64736f6c63430008170033";
```
Comment (Vid201) [inline crates/contracts/src/simulation_codes.rs:1]: remove `_STR` suffix

File: crates/contracts/src/entry_point.rs [hunk 7]
```diff
@@ -1,52 +1,109 @@
...
     },
 };
-use std::sync::Arc;
+use silius_primitives::UserOperationSigned;
+use std::{fmt::Debug, str::FromStr, sync::Arc};
 
 const UINT96_MAX: u128 = 5192296858534827628530496329220095;
 
+#[derive(EthAbiCodec, EthAbiType, Debug, Clone, PartialEq, Eq)]
+pub struct StakeInfo {
```
Comment (Vid201) [inline crates/contracts/src/entry_point.rs:34]: I think we can move this structs to primitives and create file `entrypoint.rs`

File: crates/contracts/src/gen.rs [hunk 8]
```diff
@@ -5,11 +5,43 @@ use ethers::{
...
 
-abigen!(AccountAPI, "$OUT_DIR/IAccount.sol/IAccount.json");
-abigen!(EntryPointAPI, "$OUT_DIR/IEntryPoint.sol/IEntryPoint.json");
-abigen!(PaymasterAPI, "$OUT_DIR/IPaymaster.sol/IPaymaster.json");
-abigen!(SenderCreatorAPI, "$OUT_DIR/SenderCreator.sol/SenderCreator.json");
-abigen!(StakeManagerAPI, "$OUT_DIR/IStakeManager.sol/IStakeManager.json");
+abigen!(
+    EntryPointAPI,
+    r#"[
+        struct PackedUserOperation {address sender;uint256 nonce;bytes initCode;bytes callData;bytes32 accountGasLimits;uint256 preVerificationGas;bytes32 gasFees;bytes paymasterAndData;bytes signature;}
```
Comment (Vid201) [inline crates/contracts/src/gen.rs:11]: Nice, we can get structure here!

File: crates/contracts/src/entry_point.rs [hunk 9]
```diff
@@ -80,35 +137,69 @@ impl<M: Middleware + 'static> EntryPoint<M> {
...
+    ) -> Result<SimulateValidationResult, EntryPointError> {
+        let uo = uo.into();
+        let call = self.entry_point_api.simulate_validation(uo);
+        let mut tx = call.tx;
+        tx.set_gas(100000000);
+        let call_builder = CallBuilder::new(self.eth_client.provider(), &tx);
+        let mut state = State::default();
+        state.account(self.address).code(
+            Bytes::from_str(ENTRYPOINT_SIMULATION_CODE_STR)
+                .expect("hard coded code should be correct"),
```
Comment (Vid201) [inline crates/contracts/src/entry_point.rs:185]: More meaningful panic message?

File: crates/contracts/src/entry_point.rs [hunk 10]
```diff
@@ -295,27 +393,28 @@ mod tests {
     #[tokio::test]
     #[ignore]
     async fn simulate_validation() {
-        let eth_client = Arc::new(Provider::try_from("http://127.0.0.1:8545").unwrap());
```
Comment (Vid201) [inline crates/contracts/src/entry_point.rs:298]: why change this?

**Review by Vid201 — CHANGES_REQUESTED** [review]
Good job, looks good!

I have two questinos:
- bundler spec tests are failing in the CI? Is this related to the problem you mentioned in the Telegram group?
- we need a way to build docker image for v07? we could expand the docker CI to include branch `v07`

**Comment by zsluedem** [comment]
> Good job, looks good!
> 
> I have two questinos:
> 
> * bundler spec tests are failing in the CI? Is this related to the problem you mentioned in the Telegram group?
> * we need a way to build docker image for v07? we could expand the docker CI to include branch `v07`

1. None of bundlers pass all the tests. There are some erros in the tests itself. You can see result in https://bundler-test-results.erc4337.io/v07/history/history.json. All of them have failures.
2. Yes. We should do that

File: crates/contracts/src/entry_point.rs (hunk 10 above)
Comment (zsluedem) [reply crates/contracts/src/entry_point.rs:298]: mistake

**Comment by Vid201** [comment]
> > Good job, looks good!
> > I have two questinos:
> > 
> > * bundler spec tests are failing in the CI? Is this related to the problem you mentioned in the Telegram group?
> > * we need a way to build docker image for v07? we could expand the docker CI to include branch `v07`
> 
> 1. None of bundlers pass all the tests. There are some erros in the tests itself. You can see result in https://bundler-test-results.erc4337.io/v07/history/history.json. All of them have failures.
> 2. Yes. We should do that

Hmm, ok. This PR looks good now, but it would be good to resolve these 2 things in this PR.
1. I see, but bundlers are failing only a few tests (less than 10), but in our CI (https://github.com/silius-rs/silius/actions/runs/8875058214/job/24364088083) 100 tests fail.
2. update docker CI

**Comment by zsluedem** [comment]
> > > Good job, looks good!
> > > I have two questinos:
> > > 
> > > * bundler spec tests are failing in the CI? Is this related to the problem you mentioned in the Telegram group?
> > > * we need a way to build docker image for v07? we could expand the docker CI to include branch `v07`
> > 
> > 
> > 
> > 1. None of bundlers pass all the tests. There are some erros in the tests itself. You can see result in https://bundler-test-results.erc4337.io/v07/history/history.json. All of them have failures.
> > 2. Yes. We should do that
> 
> Hmm, ok. This PR looks good now, but it would be good to resolve these 2 things in this PR.
> 
> 1. I see, but bundlers are failing only a few tests (less than 10), but in our CI (https://github.com/silius-rs/silius/actions/runs/8875058214/job/24364088083) 100 tests fail.
> 2. update docker CI

OK. Forget to update some args. Locally I got like 6 tests failing

File: .github/workflows/docker.yml [hunk 11]
```diff
@@ -2,6 +2,7 @@ on:
   push:
     branches:
       - main
+      - v07
```
Comment (Vid201) [inline .github/workflows/docker.yml:5]: Do you think it makes sense to also include v07 to the docker image name (only for v07 branch)? To differentiate between the main docker images.

File: bundler-spec-tests/docker-compose.yml [hunk 12]
```diff
@@ -2,7 +2,7 @@ version: "3"
 
 services:
   geth-dev:
-    image: ethereum/client-go:release-1.12
+    image: ethereum/client-go:release-1.13
```
Comment (Vid201) [inline bundler-spec-tests/docker-compose.yml:5]: There are some other CI errors now

**Review by Vid201 — CHANGES_REQUESTED** [review]

File: bundler-spec-tests/docker-compose.yml (hunk 12 above)
Comment (zsluedem) [reply bundler-spec-tests/docker-compose.yml:5]: Yes. debugging right now. Seems so wired though.

File: .github/workflows/docker.yml (hunk 11 above)
Comment (zsluedem) [reply .github/workflows/docker.yml:5]: Yes.

**Comment by zsluedem** [comment]
@Vid201 It should be ready now.

**Review by Vid201 — APPROVED** [review]
GJ, looks good now!

**Pull request merged**